### ISSUE TITLE
Accessibility Insights FastPass failures in Codespaces [Web]

### ISSUE THREAD

**Issue by dmgardiner25** [microsoft/vscode#130226]
Hey team, I'm running Accessibility Insights using the Edge extension for Codespaces on the web and am running into a few failures on the VS Code side of things. Here's the info of what I've been able to find from the tests:

### aria-allowed-attr
- **Failure Reason**: Ensures ARIA attributes are allowed for an element's role
- **Path**: `.codicon-explorer-view-icon`
- **Snippet**: 
```
<a class="action-label codicon codicon-explorer-view-icon" aria-label="Explorer (Ctrl+Shift+E)" style="color: rgba(255, 255, 255, 0.4);"></a>
```
- **Fix the following**: ARIA attribute cannot be used, add a role attribute or use a different element: aria-label
- **Comments**: This is on all `codicon` elements in the activity bar

### aria-required-parent
- **Failure Reason**: Ensures elements with an ARIA role that require parent roles are contained by them
- **Path**: `.icon[aria-label=""][role="tab"]:nth-child(1)`
- **Snippet**: 
```
<li class="action-item icon" role="tab" draggable="true" aria-label="" aria-expanded="false" aria-selected="false" tabindex="0" style="--insert-border-color:#ffffff;">
```
- **Fix the following**: Required ARIA parent role not present: tablist
- **Comments**: Activity bar and top of terminal buttons

### aria-required-children
- **Failure Reason**: Ensures elements with an ARIA role that require child roles contain them
- **Path**: `.setting-list-object-widget`
- **Snippet**: 
```
<div role="list" class="setting-list-object-widget setting-control-focus-target" tabindex="-1" data-focusable="true" style="height: 24px;">
  <div class="setting-list-row-header">
    <div class="setting-list-object-key">Item</div>
    <div class="setting-list-object-value">Value</div>
  </div>
</div>
```
- **Fix the following**: Required ARIA children role not present: group, listitem
- **Comments**: Item lists in settings

### aria-toggle-field-name
- **Failure Reason**: Ensures every ARIA toggle field has an accessible name
- **Path**: `.monaco-custom-checkbox`
- **Snippet**: 
```
<div title="" class="monaco-custom-checkbox codicon codicon-check setting-value-checkbox setting-control-focus-target checked" tabindex="-1" role="checkbox" aria-checked="true" aria-label="" data-focusable="true" style="border-color: transparent; color: rgb(255, 255, 255); background-color: rgba(14, 99, 156, 0.31);">
```
- **Fix the following**: One of:
     - Element does not have text that is visible to screen readers
     - aria-label attribute does not exist or is empty
     - aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
     - Element has an empty title attribute
- **Comments**: Checkboxes in the settings UI

### aria-valid-attr
- **Failure Reason**: Ensures attributes that begin with aria- are valid ARIA attributes 
- **Path**: `div[aria-label="<file name>"]`
- **Snippet**: 
```
<div draggable="true" role="tab" class="tab tab-actions-right sizing-fit has-icon" aria-label="<file name>" aria-description="" title="<file path>" data-resource-name="<file name>" aria-selected="false" tabindex="-1" style="left: auto; border-right: 1px solid rgb(37, 37, 38); background-color: rgb(45, 45, 45); color: rgba(255, 255, 255, 0.5);">
```
- **Fix the following**: Invalid ARIA attribute name: aria-description
- **Comments**: File tabs in editor

### label
- **Failure Reason**: Ensures every form element has a label
- **Path**: `.input`
- **Snippet**: 
```
<input class="input setting-control-focus-target" autocorrect="off" autocapitalize="off" spellcheck="false" type="number" wrap="off" tabindex="-1" data-focusable="true" style="background-color: inherit; color: rgb(204, 204, 204);">
```
- **Fix the following**: One of:
    - Form element does not have an implicit (wrapped) <label>
    - Form element does not have an explicit <label>
    - aria-label attribute does not exist or is empty
    - aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
    - Element has no title attribute
    - Element has no placeholder attribute
    - Element's default semantics were not overridden with role="none" or role="presentation"
- **Comments**: Input boxes in settings UI

### nested-interactive
- **Failure Reason**: Nested interactive controls are not announced by screen readers
- **Path**: `div[aria-label="<file name>"]`
- **Snippet**: 
```
<div draggable="true" role="tab" class="tab tab-actions-right sizing-fit has-icon" aria-label="<file name>" aria-description="" title="<file path>" data-resource-name="<file name>" aria-selected="false" tabindex="-1" style="left: auto; border-right: 1px solid rgb(37, 37, 38); background-color: rgb(45, 45, 45); color: rgba(255, 255, 255, 0.5);">
```
- **Fix the following**: Element has focusable descendants
- **Comments**: File tabs in editor

### image-alt
- **Failure Reason**: Ensures <img> elements have alternate text or a role of none or presentation
- **Path**: `#list_id_27_0 > .extension-list-item > .icon-container > img`
- **Snippet**: 
```
<img class="icon" src="https://vscjava.gallerycdn.vsassets.io/extensions/vscjava/vscode-java-test/0.30.1/1624863438368/Microsoft.VisualStudio.Services.Icons.Default" style="visibility: inherit;">
```
- **Fix the following**: One of:
    - Element does not have an alt attribute
    - aria-label attribute does not exist or is empty
    - aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
    - Element has no title attribute
    - Element's default semantics were not overridden with role="none" or role="presentation"
- **Comments**: Icons for extensions

### aria-allowed-attr
- **Failure Reason**: Ensures ARIA attributes are allowed for an element's role
- **Path**: `#list_id_29_0`
- **Snippet**: 
```
<div class="monaco-list-row" role="listitem" data-index="0" data-last-element="false" aria-setsize="25" aria-posinset="1" id="list_id_29_0" aria-selected="false" aria-label="Azure Account, 0.9.8, Microsoft, A common Sign-In and Subscription management extension for VS Code." data-extension-id="ms-vscode.azure-account" draggable="false" style="top: 0px; height: 62px;">
```
- **Fix the following**: ARIA attribute is not allowed: aria-selected="false"
- **Comments**: Extensions in the extension marketplace

### frame-title
- **Failure Reason**: Ensures <iframe> and <frame> elements have an accessible name
- **Path**: `.webview`
- **Snippet**: 
```
<iframe name="extensionEditor" class="webview ready" sandbox="allow-scripts allow-same-origin allow-forms allow-pointer-lock allow-downloads" allow="clipboard-read; clipboard-write;" src="https://extensionEditor.vscode-webview.net/stable/379476f0e13988d90fab105c5c19e7abc8b1dea8/out/vs/workbench/contrib/webview/browser/pre/index.html?id=extensionEditor&amp;swVersion=2&amp;extensionId=&amp;platform=browser&amp;vscode-resource-base-authority=vscode-resource.vscode-webview.net" style="border: none; width: 100%; height: 100%;">
```
- **Fix the following**: One of
    - Element has no title attribute
    - aria-label attribute does not exist or is empty
    - aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
    - Element's default semantics were not overridden with role="none" or role="presentation"
- **Comments**: Extension page content

### color-contrast
- **Failure Reason**: Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds
- **Path**: `.preview`
- **Snippet**: 
```
<span class="preview" title="Preview" style="display: inherit;">Preview</span>
```
- **Fix the following**: Element has insufficient color contrast of 2.84 (foreground color: #cccccc, background color: #d63f26, font size: 7.5pt (10px), font weight: normal). Expected contrast ratio of 4.5:1
- **Comments**: Preview tag by extension name

### aria-command-name
- **Failure Reason**: Ensures every ARIA button, link, and menuitem has an accessible name
- **Path**: `.action-dropdown-item:nth-child(8) > .monaco-dropdown > .dropdown-label > .dropdown.codicon-drop-down-button[title=""]`
- **Snippet**: 
```
<a class="action-label dropdown codicon codicon-drop-down-button extension-action label action-dropdown" role="button" aria-haspopup="true" aria-expanded="false" title=""></a>
```
- **Fix the following**: One of
    - Element does not have text that is visible to screen readers
    - aria-label attribute does not exist or is empty
    - aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
    - Element has an empty title attribute
- **Comments**: Disable and uninstall buttons on extension page

### aria-allowed-attr
- **Failure Reason**: Ensures ARIA attributes are allowed for an element's role
- **Path**: `.list_id_10`
- **Snippet**: 
```
<div class="monaco-list list_id_10 mouse-support selection-none" tabindex="0" role="list" aria-label="Tunnel View" aria-multiselectable="true" style="height: 367px;">
```
- **Fix the following**: ARIA attribute is not allowed: aria-multiselectable="true"
- **Comments**:  Codespaces port forwarding section in the terminal

### aria-allowed-attr
- **Failure Reason**: Ensures ARIA attributes are allowed for an element's role
- **Path**: `#list_id_10_0`
- **Snippet**: 
```
<div class="monaco-list-row" role="listitem" data-index="0" data-last-element="false" aria-setsize="2" aria-posinset="1" id="list_id_10_0" aria-selected="false" aria-label="Remote port localhost:2222 forwarded to local address https://dmgardiner25-bookshop-64p4p6gg25p4w-2222.githubpreview.dev/. Port has running process. Process information unavailable User Forwarded Only accessible from this machine. " draggable="false" style="top: 0px; height: 22px; line-height: 22px;">
```
- **Fix the following**: ARIA attribute is not allowed: aria-selected="false"
- **Comments**:  Codespaces port forwarding row

There might still be some that I missed on various pages but let me know if you need any more information to get these fixed or need help running the tests.

**Comment by isidorn** [comment]
@dmgardiner25 thanks for filling this. However just doing an automatic pass in my experience does not really provide any value.
Can you please for each of these issues create a separate item explaining quickly how does this actually affect a screen reader user using Codespaces?

For example here Accessibility Insighits complains about `aria-multiselectable` not being allowed, while this is in the latest aria spec and should soon be supported by all screen readers and currently is only supported by Orca.
Same for `aria-selected`. Which is part of the aria spec here https://w3c.github.io/aria/#aria-selected

Thank you and I will close this issue.
